src/dbnode/storage/index/fields_terms_iterator.go

```diff
@@ -0,0 +1,200 @@
+type fieldsAndTermsIterator interface {
+	// Next returns a bool indicating if there are any more elements.
+	Next() bool
+
+	// Current returns the current element.
+	// NB: the element returned is only valid until the subsequent call to Next().
+	Current() (field, term []byte)
+
+	// Err returns any errors encountered during iteration.
+	Err() error
+
+	// Close releases any resources held by the iterator.
+	Close() error
+
+	// Reset resets the iterator to the start iterating the given segment.
+	Reset(seg segment.Segment, opts fieldsAndTermsIteratorOpts) error
+}
+
+type fieldsAndTermsIteratorOpts struct {
+	iterateTerms bool
+	allowFn      allowFn
+}
+
+func (o fieldsAndTermsIteratorOpts) allow(f []byte) bool {
+	if o.allowFn == nil {
+		return true
+	}
+	return o.allowFn(f)
+}
+
+type allowFn func(field []byte) bool
+
+type fieldsAndTermsIter struct {
+	seg  segment.Segment
+	opts fieldsAndTermsIteratorOpts
+
+	err       error
+	fieldIter segment.FieldsIterator
+	termIter  segment.TermsIterator
+
+	current struct {
+		field []byte
+		term  []byte
+	}
+}
+
+var (
+	fieldsAndTermsIterZeroed fieldsAndTermsIter
+)
+
+var _ fieldsAndTermsIterator = &fieldsAndTermsIter{}
+
+func newFieldsAndTermsIterator(s segment.Segment, opts fieldsAndTermsIteratorOpts) (fieldsAndTermsIterator, error) {
+	iter := &fieldsAndTermsIter{}
+	err := iter.Reset(s, opts)
+	if err != nil {
+		return nil, err
+	}
+	return iter, nil
+}
+
+func (fti *fieldsAndTermsIter) Reset(s segment.Segment, opts fieldsAndTermsIteratorOpts) error {
+	*fti = fieldsAndTermsIterZeroed
+	fti.seg = s
+	fti.opts = opts
+	if s == nil {
+		return nil
+	}
+	fiter, err := s.FieldsIterable().Fields()
+	if err != nil {
+		return err
+	}
+	fti.fieldIter = fiter
+	return nil
+}
+
+func (fti *fieldsAndTermsIter) setNextField() bool {
+	fieldIter := fti.fieldIter
+	if fieldIter == nil {
+		return false
+	}
+
+	for fieldIter.Next() {
+		field := fieldIter.Current()
+		if !fti.opts.allow(field) {
+			continue
+		}
+		fti.current.field = field
+		return true
+	}
+
+	fti.err = fieldIter.Err()
+	return false
+}
+
+func (fti *fieldsAndTermsIter) setNext() bool {
+	// if only need to iterate fields
+	if !fti.opts.iterateTerms {
+		return fti.setNextField()
+	}
+
+	// if iterating both terms and fields, check if current field has another term
+	if fti.termIter != nil {
+		if fti.termIter.Next() {
+			fti.current.term, _ = fti.termIter.Current()
+			return true
+		}
+		if err := fti.termIter.Err(); err != nil {
+			fti.err = err
+			return false
+		}
+		if err := fti.termIter.Close(); err != nil {
+			fti.err = err
+			return false
+		}
+	}
+
+	// i.e. need to switch to next field
+	hasNext := fti.setNextField()
+	if !hasNext {
+		return false
+	}
+
+	// and get next term for the field
+	termsIter, err := fti.seg.TermsIterable().Terms(fti.current.field)
+	if err != nil {
+		fti.err = err
+		return false
+	}
+	fti.termIter = termsIter
+
```

Comment:
Hmm this may need to check `if !fti.opts.allow(field)` and move onto the next fieldIter if not

Reply:
nope, cause `setNextField()` does that check already.

Reply:
:+1: misread that this was calling `Next` instead